Alpha TSG client

cheapRoc opened the pull request:
At the moment this is required to demo and work on authentication within TSG. I don't believe this is something we want to merge straight into master but there are a fix or two that I can peel off for logic already in master.

The goal was to present something that would hit TSG's templates endpoint and pass along a signed HTTP request we can re-purpose for authenticating a user within TSG on behalf of CloudAPI.

WIP ~ DO NOT MERGE

cheapRoc commented:
I've opened #130 from off thread discussion around TSG endpoint configuration.

sean- commented on services/templates.go:
```diff
@@ -0,0 +1,213 @@
+	"fmt"
+	"net/http"
+	"path"
+
```
struct field AccountId should be AccountID

cheapRoc replied:
Yeah I know, we need to change `AccountId` within [joyent/triton-service-groups](http://github.com/joyent/triton-service-groups) first.

cheapRoc replied:
Actually, with joyent/triton-service-groups#13 merged I can just use struct tags for the JSON parsing and rename to `AccountID` just on the `triton-go` side.

@stack72 opened joyent/triton-service-groups#14 for the renaming.

sean- commented on services/templates.go:
```diff
@@ -0,0 +1,213 @@
+	"path"
+
+	"github.com/joyent/triton-go/client"
+	pkgerrors "github.com/pkg/errors"
```
struct field ImageId should be ImageID

sean- commented on services/templates.go:
```diff
@@ -0,0 +1,213 @@
+	}
+	respReader, err := c.client.ExecuteRequestTSG(ctx, reqInputs)
+	if respReader != nil {
+		defer respReader.Close()
```
struct field AccountId should be AccountID

sean- commented on services/templates.go:
```diff
@@ -0,0 +1,213 @@
+	if respReader != nil {
+		defer respReader.Close()
+	}
+	if err != nil {
```
struct field ImageId should be ImageID

cheapRoc commented:
For this to be useful it currently needs all open PRs under joyent/triton-service-groups as well as joyent/triton-service-groups#7 and joyent/triton-service-groups#8.

sean- commented on services/templates.go:
```diff
@@ -0,0 +1,205 @@
+
+const templatesPath = "/v1/tsg/templates"
+
+type TemplatesClient struct {
```
struct field AccountId should be AccountID

sean- commented on services/templates.go:
```diff
@@ -0,0 +1,205 @@
+
+type TemplatesClient struct {
+	client *client.Client
+}
```
struct field ImageId should be ImageID

cheapRoc commented:
Ok, we're all good on renames and JSON tags.

cheapRoc commented:
I've rebased and removed the static services URL. There is now a manual override for local development (`TRITON_TSG_URL=http://localhost:3000/` env var). There's also URL generation based on the currently selected DC (`TRITON_URL`) for when the API goes live. This transforms `https?://(.*).api.joyent.com` into `https://tsg.%s.svc.joyent.zone`. This all means you can configure a `services` client by providing the same client configuration/authentication as you would today with `triton-go` (even if TSG doesn't include auth yet).

Examples have been updated under `examples/services/` and pair with latest TSG master ([3d7c120](https://github.com/joyent/triton-service-groups/commit/3d7c120eec951e9b11c44fc36a73821f68f95801)).

cheapRoc commented:
Going to close and re-open with the branch I'm actually using for TF integration.
